components/style/properties_and_values.rs

```diff
@@ -0,0 +1,1167 @@
+            Angle(_) => true,
+            Time(_) => true,
+            TransformList(ref transform_list) => check_transform_list(transform_list),
+            CustomIdent(_) => true,
+            SpecificIdent(_) => true,
+            TokenStream(_) => true,
+        }
+    }
+
+    /// Attempts to convert this specified value to a computed value using the
+    /// given context.
+    pub fn to_computed_value(&self, context: &computed::Context) -> Result<ComputedValueItem, ()> {
+        macro_rules! compute {
+            ($_self:expr, $context:expr,
+
+             $($typ:ident),*) => {
+                match $_self {
+                    $(
+                        SpecifiedValueItem::$typ(ref value) => {
+                            Ok(ComputedValueItem::$typ(value.to_computed_value($context)))
+                        }
+                    ), *
+
+                    // Special cases.
+                    // Would put in the match syntax, but we can't have things
+                    // expand to match cases.
+
+                    SpecifiedValueItem::Url(ref url) =>
+                        ComputedUrl::from_specified(url).map(|x| ComputedValueItem::Url(x)),
+                    SpecifiedValueItem::CustomIdent(ref ident) =>
+                        Ok(ComputedValueItem::CustomIdent(ident.clone())),
+                    SpecifiedValueItem::SpecificIdent(ref ident) =>
+                        Ok(ComputedValueItem::SpecificIdent(ident.clone())),
+                    SpecifiedValueItem::TokenStream(ref stream) =>
+                        Ok(ComputedValueItem::TokenStream(stream.token_stream.clone())),
+                }
+            };
+        }
+
+        compute! {
+            *self, context,
+
+            Length, Number, Percentage, LengthPercentage, Color, Image, Integer,
+            Angle, Time, TransformList
+        }
+    }
+}
+
+/// A specified typed value.
+///
+/// A value can either be a single item or a list of items.
+#[derive(Clone, ToCss)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum SpecifiedValue {
+    /// A single specified value.
+    Item(SpecifiedValueItem),
+    /// A list of one or more specified values.
+    /// Note that we cannot have lists of <transform-lists>.
+    List(Vec<SpecifiedValueItem>),
+}
+
+/// If `list` and `x` are both `Ok(..)`, append the contents of `x` to `list`.
+/// To be used as an argument to fold to convert a sequence of `Result<T, E>`s
+/// to a `Result<Vec<T>, E>` (pulling the `Result` outside).
+fn all<T, E>(list: Result<Vec<T>, E>, x: Result<T, E>) -> Result<Vec<T>, E> {
+    list.and_then(|mut list| {
+        x.map(move |x| {
+            list.push(x);
+            list
+        })
+    })
+}
+
+impl SpecifiedValue {
+    /// Returns whether or not this specified value is computationally
+    /// independent.
+    /// See the comment on SpecifiedValueItem::is_computationally_independent.
+    pub fn is_computationally_independent(&self) -> bool {
```

Comment:
Yeah, this indeed sounds a lot like a trait that we could `derive` :)

Reply:
Certainly!